src/ion_data/ion_ord.rs

```diff
@@ -0,0 +1,262 @@
+use crate::element::{Annotations, Struct};
+use crate::{Element, IonType, Symbol};
+use std::cmp::Ordering;
+
+/// Trait used for delegating [Eq] and [PartialEq] in [IonData].
+/// Implementations of [IonOrd] must be consistent with [IonEq].
+/// Since there is no total ordering in the Ion specification, do not write any code that depends on
+/// a specific order being preserved. Only depend on the fact that a total ordering does exist.
+pub(crate) trait IonOrd {
+    // Called ion_cmp to avoid shadowing with Ord::cmp
+    fn ion_cmp(&self, other: &Self) -> Ordering;
+}
+
+impl IonOrd for IonType {
+    fn ion_cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+impl IonOrd for Annotations {
+    fn ion_cmp(&self, other: &Self) -> Ordering {
+        self.iter().cmp(other.iter())
+    }
+}
+
+impl IonOrd for f64 {
+    fn ion_cmp(&self, other: &Self) -> Ordering {
+        self.total_cmp(other)
+    }
+}
```

Comment:
It surprised me that `IonOrd` was not implemented (even trivially) for all the underlying value types.  I understand why now looking at `element::Value`, but I think `IonOrd` should be built on top of `IonOrd` and not a combination of `IonOrd` and `PartialOrd`.  If we don't like this, I think this really should be documented because what is unusual about this is that some of the underlying value types work with `IonData` and some don't.  E.g., I *think* I can  `IonData<f64>` but not `IonData<i64>` if I read this all correctly (and I may not be).

I was re-looking at `IonEq` and I think this is the source of the asymmetry.  Ideally we'd have blanket implementations and specializations for when `IonEq` doesn't apply (but specialization is not stable AFAIK), so I think it best to just comment the intention for `IonOrd`/`IonEq` **or** implement it for all of the constituent "value" types underlying the value enum.

Reply:
I can add trivial `IonEq` and `IonOrd` for all of the remaining value types, if you think that is also an appropriate solution.